core/profiling/eventpipe/src/CorProfiler.cpp

```diff
@@ -0,0 +1,781 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "CorProfiler.h"
+#include "corhlpr.h"
+#include "CComPtr.h"
+#include "profiler_pal.h"
+#include <string>
+#include <sstream>
+#include <assert.h>
+#include <cstdlib>
+#include <ctime>
+
+using std::shared_ptr;
+using std::vector;
+using std::mutex;
+using std::lock_guard;
+using std::map;
+using std::thread;
+
+CorProfiler::CorProfiler() :
+    _pCorProfilerInfo12(),
+    _session(),
+    _provider(),
+    _allTypesEvent(),
+    _providerNameCache(),
+    _cacheLock(),
+    _metadataCache(),
+    _refCount(0)
+{
+
+}
+
+CorProfiler::~CorProfiler()
+{
+    if (this->_pCorProfilerInfo12 != nullptr)
+    {
+        this->_pCorProfilerInfo12->Release();
+        this->_pCorProfilerInfo12 = nullptr;
+    }
+}
+
+HRESULT STDMETHODCALLTYPE CorProfiler::Initialize(IUnknown *pICorProfilerInfoUnk)
+{
+    HRESULT hr = S_OK;
+    if (FAILED(hr = pICorProfilerInfoUnk->QueryInterface(__uuidof(ICorProfilerInfo12), (void **)&_pCorProfilerInfo12)))
+    {
+        printf("FAIL: failed to QI for ICorProfilerInfo12.\n");
+        return hr;
+    }
+
+    if (FAILED(hr = _pCorProfilerInfo12->SetEventMask2(COR_PRF_MONITOR_JIT_COMPILATION
+                                                        | COR_PRF_DISABLE_ALL_NGEN_IMAGES,
+                                                       COR_PRF_HIGH_MONITOR_EVENT_PIPE)))
+    {
+        printf("FAIL: ICorProfilerInfo::SetEventMask2() failed hr=0x%x\n", hr);
+        return hr;
+    }
+    
+    // Write an event
+    if (FAILED(hr = DefineEvent()) || FAILED(hr = WriteEvent()))
+    {
+        return hr;
+    }
+
+    // Start a session for listening
+    if (FAILED(hr = StartSession()))
+    {
+        return hr;
+    }
+
+    return S_OK;
+}
+
+HRESULT CorProfiler::DefineEvent()
```

Comment:
I think I want to stay away from making this an all inclusive tutorial on eventing. I agree that more comments than what I currently have would be nice, but I think the comments on this sample should be related to things that are specific to how the profiler interacts with eventpipe.

I.e. I things like what is an event and how to version them should be in a standalone doc somewhere, then we can link from this sample to the doc. At a cursory glance I don't see that doc anywhere, so that seems like a todo on our team.

Things like why is the sample creating these events should be explained here.